### PR TITLE
Make CI fail on warnings

### PR DESCRIPTION
**Pull request by fsimonis**
## Main changes of this PR

Makes the CI fail on warnings.

## Motivation and additional information

This should prevent future warning creeping into the code base

## Author's checklist

* [ ] I added a changelog file with this PR number in `docs/changelog/` if there are noteworthy changes.
* [ ] I ran `tools/formatting/check-format` and everything is formatted correctly.
* [ ] I sticked to C++14 features.
* [ ] I sticked to CMake version 3.10.
* [ ] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?
* [ ] (more questions/tasks)

### CODE REVIEW

File: src/cplscheme/BaseCouplingScheme.hpp [hunk 1]
```diff
@@ -27,7 +27,7 @@ class TXTTableWriter;
 } // namespace io
 
 namespace cplscheme {
-struct CouplingData;
+class CouplingData;
```
Comment (uekerman) [inline src/cplscheme/BaseCouplingScheme.hpp:30]: @BenjaminRodenberg already did the same change in #1004 
Just fyi

Comment (fsimonis) [reply src/cplscheme/BaseCouplingScheme.hpp:30]: This shouldn't lead to conflicts as the change is identical.

**Comment by fsimonis** [comment]
I spend some time pondering on this and I don't thing the hassle of `-Werror`  is worth it until we have a good way to bundle assertions with required code-blocks. Otherwise we will constantly run into false-positives for cases like this:

```
auto x = somethingBigProducingX();
PRECICE_ASSERT( something using x );
```

I'll migrate changes over to develop and close this PR.

**Comment by fsimonis** [comment]
I moved relevant changes into separate PRs.

**Pull request closed without merging**